### PR TITLE
Validate infrastructure related configuration only on start workspace

### PR DESCRIPTION
**Pull request by gazarenkov**
### Motivation: 

Currently, we validate recipe on both workspace create/update and start.
It is not convenient policy since in this case we consider workspace/stack storage to have only workspaces with ONLY the same supported type(s) of infrastructure as supported by runtime.
It is not convenient for portable workspaces since:

- it does not allow to have several environments of different recipe types inside a workspace 
- we can consider keeping shared registry among different infrastructures
- we can consider using environment of one type to be imported and used as a template for other, "supported" type on demand

### Side effect:

Workspace storage may contain incorrect recipes but that's a reasonable drawback since it will be validated before start

### What does this PR do?

Logic of recipe validating moved from "generic" validation part to start Workspace

### CODE REVIEW

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java [hunk 1]
```diff
@@ -343,7 +344,7 @@ public void stopWorkspace(String workspaceId, Map<String, String> options)
             });
   }
 
-  public Set<String> getSupportedRecipes() {
+  Set<String> getSupportedRecipes() {
```
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java]: Could you explain why this method should be package private but not public?

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceValidator.java [hunk 2]
```diff
@@ -94,12 +85,6 @@ public void validateConfig(WorkspaceConfig config)
           environment.getMachines().entrySet()) {
         validateMachine(machineEntry.getKey(), machineEntry.getValue());
       }
-
-      try {
-        runtimes.validate(environment);
```
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceValidator.java:99]: Java doc should be updated along with these changes https://github.com/eclipse/che/pull/10385/files#diff-5f46ddd221720706a3210208652a0f1eR57

**Review by sleshchenko — COMMENTED** [review]
> it does not allow to have several environments of different recipe types inside a workspace

Let me clarify a bit. Now it is possible to have several environments of different recipe types inside a workspace. BUT it is not possible to have environments which have recipes that are currently not supported. At the same time, at least one supported environment is enough to be able to start a workspace.

> we can consider using environment of one type to be imported and used as a template for other, "supported" type on demand

Could you elaborate more on what do you mean using an environment as `template` for other?

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java (hunk 1 above)
Comment (gazarenkov) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java]: Well, there are no other reasons than IntelliJ suggestion and absence of javadoc (I thought it is not important enough to be public )) ). If you think it is useful, I do not mind to revert and add javadocs as well.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceValidator.java (hunk 2 above)
Comment (gazarenkov) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceValidator.java:99]: done

**Comment by gazarenkov** [comment]
> we can consider using environment of one type to be imported and used as a template for other, "supported" type on demand
> 
> Could you elaborate more on what do you mean using an environment as template for other?

Sure, nothing special in fact:
1/ You have workspace with recipe of type1 in one instance (supporting type1)
2/ You want to make it work in other instance which supports recipe of type2
3/ You export this workspace and import to other instance with no supportability problem (due to this update)
4/ You modify it as you want to make it support type2 and test

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceValidator.java [hunk 3]
```diff
@@ -49,26 +46,16 @@
   private static final Pattern VOLUME_NAME = Pattern.compile("[a-z][a-z0-9]{1,18}");
   private static final Pattern VOLUME_PATH = Pattern.compile("/.+");
 
-  private final WorkspaceRuntimes runtimes;
```
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceValidator.java:52]: It does not affect anything but please remove `WorkspaceRuntimes` from `WorkspaceValidator` too
https://github.com/eclipse/che/blob/2fb979bbe1cf2fc5f4179c2440bfd4de9be73749/wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/WorkspaceValidatorTest.java#L46

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java [hunk 4]
```diff
@@ -343,6 +344,7 @@ public void stopWorkspace(String workspaceId, Map<String, String> options)
             });
   }
 
+  /** @return set of supported recipe types */
```
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java]: According to [Google Java Code Style](https://google.github.io/styleguide/javaguide.html#s7.2-summary-fragment) it makes sense to reformat this java doc as 
```
/** Returns set of supported recipe types */
```
Up to you change it or not

**Review by sleshchenko — APPROVED** [review]
LGTM

**Review by skabashnyuk — APPROVED** [review]

**Comment by riuvshin** [comment]
ci-test build report:
[Build details](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/471/)
[Test report](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/471//Selenium_tests_report)
[selenium tests report data](http://tests.codenvycorp.com/selenium_test_report_data/10385)
docker image: eclipseche/che-server:10385
https://github.com/orgs/eclipse/teams/eclipse-che-qa please check this report.

**Comment by Ohrimenko1988** [comment]
In the functionality which is covered by selenium tests - new regression was not found.

**Pull request merged**